### ISSUE TITLE
Result of `expand_db_html` for a mixture of internal / external links is malformed

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#11957]
### Issue Summary

Running `expand_db_html` on rich text input containing a mix of internal and external links results in malformed output, presumably as a consequence of #5875.

### Steps to Reproduce

Within `./manage.py shell` on a bakerydemo instance:

```python
from wagtail.rich_text import expand_db_html
rich_text = '<p data-block-key="9l01v"><a href="https://wagtail.org/">External link 1</a></p>'\
            '<p data-block-key="b1mdf"><a linktype="page" id="3">Internal link</a></p>'\
            '<p data-block-key="1hsrl"><a href="https://wagtail.org/">External link 2</a></p>'
expand_db_html(rich_text)
```

results in

```
'<p data-block-key="9l01v"><a href="/breads/">External link 1</a></p><p data-block-key="b1mdf"><a linktype="page" id="3">Internal link</a></p><p data-block-key="1hsrl"><a href="https://wagtail.org/">External link 2</a></p>'
```

Note that the URL `/breads/` (corresponding to page ID 3) has been substituted on external link 1 instead of the internal link as expected.

Additionally, with [the NoFollowExternalLinkHandler example from the Wagtail docs](https://docs.wagtail.org/en/stable/extending/rich_text_internals.html#FeatureRegistry.register_link_type) defined in `base/wagtail_hooks.py`:

```python
from wagtail.rich_text import LinkHandler
from django.utils.html import escape
from wagtail import hooks


class NoFollowExternalLinkHandler(LinkHandler):
    identifier = 'external'

    @classmethod
    def expand_db_attributes(cls, attrs):
        href = attrs["href"]
        return '<a href="%s" rel="nofollow">' % escape(href)


#@hooks.register('register_rich_text_features')
def register_external_link(features):
    features.register_link_type(NoFollowExternalLinkHandler)
```

we get the output

```
'<p data-block-key="9l01v"><a href="https://wagtail.org/" rel="nofollow">External link 1</a></p><p data-block-key="b1mdf"><a linktype="pa<a href="/breads/">p><p data-block-key="1hsrl"><a href="https://wagtail.org/" rel="nofollow">External link 2</a></p>'
```

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

-   Python version: Run `python --version`.
-   Django version: Look in your requirements.txt, or run `pip show django | grep Version`.
-   Wagtail version: Look at the bottom of the Settings menu in the Wagtail admin, or run `pip show wagtail | grep Version:`.
-   Browser version: You can use https://www.whatsmybrowser.org/ to find this out.

### Working on this

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

**Comment by chosak** [comment]
@gasman thanks for finding this. I think the bug is in these lines:

https://github.com/wagtail/wagtail/blob/4f5ffa85b660bb76c5db43e72d5b6f37b884fb06/wagtail/rich_text/rewriters.py#L56-L58

which don't appropriately handle mixed-type links where one type doesn't need replacements. I can open a PR that adds tests for these cases.

**Comment by gasman** [comment]
The second issue can be replicated without a custom link handler:

```python
expand_db_html('<a linktype="page" id="3">page</a><a linktype="document" id="1">document</a><a linktype="page" id="3">page</a>')
```

It occurs because [the replacement logic in `TagRewriter.call`](https://github.com/wagtail/wagtail/blob/8ddf472e93710a68ec4a44fbc853239c53d25c60/wagtail/rich_text/rewriters.py#L55-L65) keeps track of the offset to apply to `match.start()` / `match.end()` by adding up all of the previous length differences from previously-performed rewrites. This assumes that we'll always be moving forward through the string, which isn't the case when more than one linktype is present.

**Comment by gasman** [comment]
Fixed in #11959
